go/vt/vttablet/tabletmanager/rpc_replication.go

```diff
@@ -567,15 +576,11 @@ func (agent *ActionAgent) setMasterLocked(ctx context.Context, parentAlias *topo
 	// Being sent SetMaster means another MASTER has been successfully promoted,
 	// so we convert to REPLICA first, since we want to do it even if other
 	// steps fail below.
-	_, err = agent.TopoServer.UpdateTabletFields(ctx, agent.TabletAlias, func(tablet *topodatapb.Tablet) error {
-		if tablet.Type == topodatapb.TabletType_MASTER {
-			tablet.Type = topodatapb.TabletType_REPLICA
-			return nil
+	if agent.Tablet().Type == topodatapb.TabletType_MASTER {
```

Comment:
I think we should just go ahead and always call `ChangeType`. If we add the `NoUpdateNeeded` optimization in ChangeType as discussed in the other comments, this will do the right thing.

The goal of this update is to fix topo no matter what. To be idempotent, it was important that we don't skip any step A (fix topo) just because we found that some other step B (change agent.Tablet().Type) had already succeeded. We should only skip A if we directly confirm A is already in the desired state (e.g. if ChangeType determines that no update is needed).

Reply:
Here we really do want to check the type first, because we could be calling SetMaster on an RDONLY tablet.

Reply:
I agree. It didn't feel right to check the local tablet type instead of what is in topo. I will change it back.